### PR TITLE
(feat codewhisperer): inline tutorial

### PR DESCRIPTION
**Pull request by Will-ShaoHua**
## Solution
NOTE: Inline tutorial is targeting  *New users* ONLY
Existing users won't be able to see the new UI

`LineAnnotationController` is a component for the following purpose
- manage the current state/step of the tutorial user is at
- update the UI if needed
- Each state/step of the tutorial has its own exit criteria, only exit criteria is met will the tutorial proceed to next state/step
- Users are able to press "ESC" key to dismiss the entire tutorial at anytime

## Walk through

case 1->1a

https://github.com/aws/aws-toolkit-vscode/assets/96078566/fea65837-3b15-4f51-9916-31dd811f5e77

case 1->2

https://github.com/aws/aws-toolkit-vscode/assets/96078566/c4180765-3b20-44d5-a397-bf15b324229f

case 2->3

https://github.com/aws/aws-toolkit-vscode/assets/96078566/91a3e5f8-22ae-42ec-b45e-d0f5317187aa

case 3 -> done

30s timeout

https://github.com/aws/aws-toolkit-vscode/assets/96078566/3f061332-e391-4eff-830e-1bb374b193f6

click statusbar

https://github.com/aws/aws-toolkit-vscode/assets/96078566/98c970ae-c10e-4b4f-b44a-935d52b047bc

Sign-out

https://github.com/aws/aws-toolkit-vscode/assets/96078566/9c177f32-8938-4e08-86ca-093d71de44e0

Sign-in

https://github.com/aws/aws-toolkit-vscode/assets/96078566/c6be26b6-e21a-4cc3-9ec9-492129fa8013

Existing users

https://github.com/aws/aws-toolkit-vscode/assets/96078566/5a0cc977-b2a0-4f60-9c18-396c2cdcc872

New users

https://github.com/aws/aws-toolkit-vscode/assets/96078566/76e3b37d-ebdf-4ad0-95d1-e78a1d2c3350

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by justinmk3** [comment]
Will this use the vscode walkthrough feature: https://code.visualstudio.com/api/ux-guidelines/walkthroughs ?

We had previous walkthroughs but they were removed:
https://github.com/aws/aws-toolkit-vscode/pull/3630

**Comment by Will-ShaoHua** [comment]
> Will this use the vscode walkthrough feature: https://code.visualstudio.com/api/ux-guidelines/walkthroughs ?
> 
> We had previous walkthroughs but they were removed: #3630

nahh, it won't use the feature you mentioned @justinmk3. let put a screen recording later in this PR, we're only using the inline hint (decoration) to provide new users some general guidance

File: packages/core/package.json [hunk 1]
```diff
@@ -3894,6 +3894,12 @@
                 "mac": "escape",
                 "when": "inlineSuggestionVisible && !editorReadonly && aws.codewhisperer.connected || isCloud9 && suggestWidgetVisible && !editorReadonly"
             },
+            {
+                "command": "aws.codeWhisperer.dismissTutorial",
+                "key": "escape",
+                "mac": "escape",
+                "when": "aws.codewhisperer.tutorial.workInProgress && !inlineSuggestionVisible && !suggestWidgetVisible"
+            },
```
Comment (Will-ShaoHua) [inline packages/core/package.json:3902]: Command for users to press [esc] key to dismiss/escape from the tutorial. It won't be shown again afterward.

**Review by andrewyuq — APPROVED** [review]

File: packages/core/src/codewhisperer/service/recommendationHandler.ts [hunk 2]
```diff
@@ -62,6 +62,8 @@ const nextCommand = Commands.declare('editor.action.inlineSuggest.showNext', ()
 
 const rejectCommand = Commands.declare('aws.codeWhisperer.rejectCodeSuggestion', () => async () => {
     RecommendationHandler.instance.reportUserDecisions(-1)
+
+    await Commands.tryExecute('aws.codewhisperer.refreshAnnotation')
```
Comment (Will-ShaoHua) [inline packages/core/src/codewhisperer/service/recommendationHandler.ts:66]: I was trying
both
(1.) to use declared and exported command `aws.codewhisperer.refreshAnnotation`
and 
(2.)
```
Container.instance.lineAnnotationController.refresh()
```

however both encountered "circular dependency" issue, thus having to set back to this approach which may not be preferred

File: packages/core/src/codewhisperer/activation.ts [hunk 3]
```diff
@@ -220,7 +226,7 @@ export async function activate(context: ExtContext): Promise<void> {
         /**
          * On recommendation acceptance
          */
-        acceptSuggestion.register(context),
+        acceptSuggestion.register(context, container),
```
Comment (Will-ShaoHua) [inline packages/core/src/codewhisperer/activation.ts:229]: any advice on this? 
another way to do it is just call `Container.instance` within the commands

**Comment by Will-ShaoHua** [comment]
for example commit https://github.com/aws/aws-toolkit-vscode/pull/4539/commits/1441342cbd3bd53a5c09c4e95969414779798231 will cause circular dependency
```
 ERROR in Circular dependency detected:
  src/codewhisperer/service/inlineCompletionService.ts -> src/codewhisperer/ui/statusBarMenu.ts -> src/codewhisperer/service/serviceContainer.ts -> src/codewhisperer/views/lineAnnotationController.ts -> src/codewhisperer/service/recommendationService.ts -> src/codewhisperer/service/inlineCompletionService.ts

  ERROR in Circular dependency detected:
  src/codewhisperer/service/recommendationService.ts -> src/codewhisperer/service/inlineCompletionService.ts -> src/codewhisperer/ui/statusBarMenu.ts -> src/codewhisperer/service/serviceContainer.ts -> src/codewhisperer/views/lineAnnotationController.ts -> src/codewhisperer/service/recommendationService.ts

  ERROR in Circular dependency detected:
  src/codewhisperer/service/serviceContainer.ts -> src/codewhisperer/views/lineAnnotationController.ts -> src/codewhisperer/service/recommendationService.ts -> src/codewhisperer/service/inlineCompletionService.ts -> src/codewhisperer/ui/statusBarMenu.ts -> src/codewhisperer/service/serviceContainer.ts

  ERROR in Circular dependency detected:
  src/codewhisperer/ui/statusBarMenu.ts -> src/codewhisperer/service/serviceContainer.ts -> src/codewhisperer/views/lineAnnotationController.ts -> src/codewhisperer/service/recommendationService.ts -> src/codewhisperer/service/inlineCompletionService.ts -> src/codewhisperer/ui/statusBarMenu.ts

  ERROR in Circular dependency detected:
  src/codewhisperer/views/activeStateController.ts -> src/codewhisperer/service/recommendationService.ts -> src/codewhisperer/service/inlineCompletionService.ts -> src/codewhisperer/ui/statusBarMenu.ts -> src/codewhisperer/service/serviceContainer.ts -> src/codewhisperer/views/activeStateController.ts

  ERROR in Circular dependency detected:
  src/codewhisperer/views/lineAnnotationController.ts -> src/codewhisperer/service/recommendationService.ts -> src/codewhisperer/service/inlineCompletionService.ts -> src/codewhisperer/ui/statusBarMenu.ts -> src/codewhisperer/service/serviceContainer.ts -> src/codewhisperer/views/lineAnnotationController.ts

  web (webpack 5.83.0) compiled with 6 errors and 14 warnings in 37672 ms
npm ERR! Lifecycle script `compile` failed with error: 
npm ERR! Error: command failed 
npm ERR!   in workspace: aws-core-vscode@1.0.0 
```

**Comment by Will-ShaoHua** [comment]
fixing circular dependency...

File: packages/toolkit/package.json [hunk 4]
```diff
@@ -1,82 +0,0 @@
-{
```
Comment (Will-ShaoHua) [inline packages/toolkit/package.json:1]: will revert

**Comment by Will-ShaoHua** [comment]
circular dep issue is resolved, will upload screen recordings when testing is done

**Comment by Will-ShaoHua** [comment]
attached recording of different scenarios

**Review by nkomonen-amazon — APPROVED** [review]

**Comment by Will-ShaoHua** [comment]
mac minimum failed with, i feel it's irrelevant to this PR 
```
TypeError: Cannot read properties of undefined (reading 'asAbsolutePath')
    at resolveIconId (/Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/packages/core/src/shared/icons.ts:90:33)
    at /Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/packages/core/src/shared/utilities/functionUtils.ts:77:64
    at createFeedbackNode (/Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/packages/core/src/codewhisperer/ui/codeWhispererNodes.ts:208:26)
    at Suite.<anonymous> (/Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/packages/core/src/test/codewhisperer/commands/basicCommands.test.ts:301:49)
    at Object.create (/Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/node_modules/mocha/lib/interfaces/common.js:148:19)
    at context.describe.context.context (/Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/node_modules/mocha/lib/interfaces/bdd.js:42:27)
    at Suite.<anonymous> (/Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/packages/core/src/test/codewhisperer/commands/basicCommands.test.ts:300:5)
    at Object.create (/Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/node_modules/mocha/lib/interfaces/common.js:148:19)
    at context.describe.context.context (/Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/node_modules/mocha/lib/interfaces/bdd.js:42:27)
    at Object.<anonymous> (/Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/packages/core/src/test/codewhisperer/commands/basicCommands.test.ts:55:1)
    at Module._compile (node:internal/modules/cjs/loader:1163:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1216:10)
    at Module.load (node:internal/modules/cjs/loader:1035:32)
    at Module._load (node:internal/modules/cjs/loader:876:12)
    at Function.c._load (node:electron/js2c/asar_bundle:5:13343)
    at Function.s._load (/Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/packages/core/.vscode-test/vscode-darwin-1.68.0/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:102:32307)
    at Function.b._load (/Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/packages/core/.vscode-test/vscode-darwin-1.68.0/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:102:28750)
    at Function._._load (/Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/packages/core/.vscode-test/vscode-darwin-1.68.0/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:97:61578)
    at Module.require (node:internal/modules/cjs/loader:1059:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at /Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/node_modules/mocha/lib/mocha.js:414:36
    at Array.forEach (<anonymous>)
    at Mocha.loadFiles (/Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/node_modules/mocha/lib/mocha.js:411:14)
    at Mocha.run (/Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/node_modules/mocha/lib/mocha.js:969:10)
    at /Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/packages/core/src/test/testRunner.ts:94:19
    at new Promise (<anonymous>)
    at runMocha (/Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/packages/core/src/test/testRunner.ts:93:16)
    at runTests (/Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/packages/core/src/test/testRunner.ts:118:11)
```

**Comment by justinmk3** [comment]
> Looks like 'vscode-thumbsup' couldn't be found and return undefined?

We should never fail in that case. Can we fix `resolveIconId` (or related) so that it returns a generic icon if the requested icon couldn't be found?

File: packages/core/src/codewhisperer/views/lineAnnotationController.ts [hunk 5]
```diff
@@ -435,11 +426,13 @@ export class LineAnnotationController implements vscode.Disposable {
...
-            source: source,
-            recommendationCount: e?.response?.recommendationCount ?? 0,
-        })
+        let nextState: AnnotationState | undefined
+
+        if (force) {
+            nextState = this._currentState.nextState(source, true)
+        } else {
+            nextState = this._currentState.nextState(source, false)
+        }
```
Comment (justinmk3) [inline packages/core/src/codewhisperer/views/lineAnnotationController.ts:433]: nextState = this._currentState.nextState(source, force)

Comment (justinmk3) [inline packages/core/src/codewhisperer/views/lineAnnotationController.ts:433]: const nextState = this._currentState.nextState(source, force)

Comment (nkomonen-amazon) [reply packages/core/src/codewhisperer/views/lineAnnotationController.ts:433]: Addressed here: https://github.com/aws/aws-toolkit-vscode/pull/4554

**Pull request merged**